pkg/karmadactl/rebalance/rebalance.go

```diff
@@ -0,0 +1,253 @@
+	}
+
+	// 3. rebalance object
+	return o.rebalanceObject(karmadaClient, obj)
+}
+
+// getObjInfo get obj info in member cluster
+func (o *CommandRebalanceOption) getObjInfo(f cmdutil.Factory, args []string) (*unstructured.Unstructured, error) {
+	r := f.NewBuilder().
+		Unstructured().
+		NamespaceParam(o.Namespace).
+		FilenameParam(false, &o.FilenameOptions).
+		RequestChunksOf(500).
+		ResourceTypeOrNameArgs(true, args...).
+		ContinueOnError().
+		Latest().
+		Flatten().
+		Do()
+
+	r.IgnoreErrors(apierrors.IsNotFound)
+
+	infos, err := r.Infos()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(infos) == 0 {
+		return nil, fmt.Errorf("the %s or %s don't exist in Karmada", args[0], args[1])
+	}
+
+	return infos[0].Object.(*unstructured.Unstructured), nil
+}
+
+func (o *CommandRebalanceOption) rebalanceObject(cli karmadaclientset.Interface, obj *unstructured.Unstructured) error {
+	if o.DryRun {
+		fmt.Printf("try to rebalance resource %s %s %s\n", obj.GetKind(), obj.GetNamespace(), obj.GetName())
+		return nil
+	}
+
+	rebalancer := &appsv1alpha1.WorkloadRebalancer{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      string(uuid.NewUUID()),
+			Namespace: obj.GetNamespace(),
+			Labels:    map[string]string{rebalancerByCtl: "true"},
+		},
+		Spec: appsv1alpha1.WorkloadRebalancerSpec{
+			Workloads: []appsv1alpha1.ObjectReference{{
+				APIVersion: obj.GetAPIVersion(),
+				Kind:       obj.GetKind(),
+				Name:       obj.GetName(),
+				Namespace:  obj.GetNamespace(),
+			}},
+			TTLSecondsAfterFinished: ptr.To[int32](30 * 60), // default ttl 30 minutes
+		},
+	}
+
+	_, err := cli.AppsV1alpha1().WorkloadRebalancers().Create(context.TODO(), rebalancer, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to trigger rebalance when creating WorkloadRebalancer: %w", err)
+	}
+
+	err = wait.PollUntilContextTimeout(context.TODO(), checkInterval, checkTimeout, false,
+		func(ctx context.Context) (done bool, err error) {
+			// 1. get WorkloadRebalancer
+			rebalancerGet, err := cli.AppsV1alpha1().WorkloadRebalancers().Get(ctx, rebalancer.Name, metav1.GetOptions{})
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					return false, fmt.Errorf("unknown rebalance result, WorkloadRebalancer/%s seems unexpected deleted", rebalancer.Name)
+				}
+				return false, err
+			}
+
+			// 2. wait for expect result: observedWorkload's result is successful
+			if len(rebalancerGet.Status.ObservedWorkloads) == 0 {
+				// haven't been reconciled, retry
+				return false, nil
+			}
+			observedWorkload := rebalancerGet.Status.ObservedWorkloads[0]
+			if observedWorkload.Result == "" {
+				// haven't finished, retry
+				return false, nil
+			}
+			if observedWorkload.Result == appsv1alpha1.RebalanceFailed {
+				// already failed, return
+				return true, fmt.Errorf("rebalance failed due to %s", observedWorkload.Reason)
+			}
+
+			// 3. if rebalance successfully, directly remove rebalancer.
```

Comment:
After rebalance succeeds, should the rebalancer be deleted directly? Users may need the rebalance manifest to trace the entire process.

Reply:
Rebalancer itself is a useful piece of process information. Its presence indicates that the entire process is carried out through the rebalancer. Users can verify that the rebalancer is generated as expected by reviewing its manifest, such as the Workload field.